backend/main.go

```diff
@@ -0,0 +1,167 @@
+package main
+
+import (
+	"encoding/csv"
+	"encoding/json"
+	"fmt"
+	"io"
+	"log"
+	"os"
+	"strconv"
+	"time"
+
+	"foxygo.at/s/errs"
+	"github.com/tidwall/pretty"
+)
+
+// ErrInvalidInput is a sentinel error for invalid input in CSV file.
+var ErrInvalidInput = fmt.Errorf("invalid input")
+
+// Data is the top level structure holding all covid time series data
+// per country. For memory efficiency dates are only listed once.
+type Data struct {
+	Dates     []time.Time         `json:"dates"`
+	Countries map[string]*Country `json:"countries"`
+}
+
+// Country is a sub-struct of Data holding all time series data for a
+// specific country some meta-data such as population.
+type Country struct {
+	Name       string  `json:"name"`
+	Lat        float64 `json:"lat"`
+	Lng        float64 `json:"lng"`
+	Population int64   `json:"population"`
+	// per date new, not cumulative, confirmed cases/recovered/dead:
+	Confirmed []int `json:"confirmed"`
+	Recovered []int `json:"recovered"`
+	Dead      []int `json:"dead"`
+}
+
+func main() {
+	data, err := parse(os.Stdin)
+	if err != nil {
+		log.Fatal(err)
+	}
+	b, err := json.Marshal(data)
+	if err != nil {
+		log.Fatal(err)
+	}
+	opts := &pretty.Options{Width: 10000, Prefix: "", Indent: "  "}
+	fmt.Println(string(pretty.PrettyOptions(b, opts)))
+}
+
+func parse(r io.Reader) (*Data, error) {
+	csvr := csv.NewReader(r)
+	csvr.FieldsPerRecord = 8
+	csvr.ReuseRecord = true
+	// Date,       Country/Region,Province/State,Lat,  Long, Confirmed,Recovered,Deaths
+	// 2020-01-22, Afghanistan,                 ,33.0, 65.0, 0,        0,        0
+	rows, err := csvr.ReadAll()
+	if err != nil {
+		return nil, err
+	}
+	rows = rows[1:] // skip header
+	dates, err := parseDates(rows)
+	if err != nil {
+		return nil, err
+	}
+	countries, err := parseCountries(rows, dates)
+	if err != nil {
+		return nil, err
+	}
+	return &Data{Dates: dates, Countries: countries}, nil
+}
+
+func parseDates(rows [][]string) ([]time.Time, error) {
+	d := map[string]bool{}
+	dates := []time.Time{}
+	for _, row := range rows {
+		if d[row[0]] {
+			break
+		}
+		d[row[0]] = true
+		date, err := time.Parse("2006-01-02", row[0])
+		if err != nil {
+			return nil, err
+		}
+		dates = append(dates, date)
+	}
+	return dates, nil
+}
+
+func parseCountries(rows [][]string, dates []time.Time) (map[string]*Country, error) {
+	countries := map[string]*Country{}
+	dateIdx := 0
+	dateCnt := len(dates)
+	for _, row := range rows {
+		country, province := row[1], row[2]
+		if province != "" {
```

Comment:
This looks like it skips countries with provinces. At least for Australia it does - there are no Australian data with an empty province in the CSV. Same with Canada too.

Reply:
yes, shame on me for being so sloppy.
i have addressed this.